lib/collaborative_filtering.py

```diff
@@ -2,3 +2,212 @@
+        @param (bool) verbose if True, tracing will be printed
+        """
+        self.ratings = ratings
+        self.n_users, self.n_items = ratings.shape
+        self.n_factors = config['n_factors']
+        self._lambda = config['_lambda']
+        self.evaluator = evaluator
+        self._v = verbose
+
+    def process(self):
+        """
+        Start processing the data.
+        """
+        pass
+
+    def set_config(self, config):
+        """
+        The function sets the config of the uv_decomposition algorithm
+        @param (dict) config hyperparameters of the recommender, contains
+        _lambda and n_factors
+        """
+        self.n_factors = config['n_factors']
+        self._lambda = config['_lambda']
+
+    def split(self):
+        test = numpy.zeros(self.ratings.shape)
+        train = self.ratings.copy()
+        # TODO split in a more intelligent way
+        for user in range(self.ratings.shape[0]):
+            test_ratings = numpy.random.choice(self.ratings[user, :].nonzero()[0],
+                                               size=10)
+            train[user, test_ratings] = 0.
+            test[user, test_ratings] = self.ratings[user, test_ratings]
+        assert(numpy.all((train * test) == 0))
+        self.ratings = train
+        return train, test
+
+    def als_step(self,
+                 latent_vectors,
+                 fixed_vecs,
+                 ratings,
+                 _lambda,
+                 type='user'):
+        """
+        The function computes only one step in the ALS algorithm
+        @param latent_vectors (ndarray) the vector to be optimized
+        @param fixed_vecs (ndarray) the vector to be fixed
+        @param ratings (ndarray) ratings that will be used to optimize latent * fixed
+        @param _lambda (float) reguralization parameter
+        @param type (string) either user or item.
+        """
+        if type == 'user':
+            # Precompute
+            YTY = fixed_vecs.T.dot(fixed_vecs)
+            lambdaI = numpy.eye(YTY.shape[0]) * _lambda
+
+            for u in range(latent_vectors.shape[0]):
+                latent_vectors[u, :] = solve((YTY + lambdaI),
+                                             ratings[u, :].dot(fixed_vecs))
+        elif type == 'item':
+            # Precompute
+            XTX = fixed_vecs.T.dot(fixed_vecs)
+            lambdaI = numpy.eye(XTX.shape[0]) * _lambda
+            for i in range(latent_vectors.shape[0]):
+                latent_vectors[i, :] = solve((XTX + lambdaI),
+                                             ratings[:, i].T.dot(fixed_vecs))
+        return latent_vectors
+
+    def train(self, n_iter=5):
+        """
+        Train model for n_iter iterations from scratch.
+        @param n_iter (int) number of iterations
+        """
+        self.user_vecs = numpy.random.random((self.n_users, self.n_factors))
+        self.item_vecs = numpy.random.random((self.n_items, self.n_factors))
```

Comment:
You can add the option somehow to get the theta from a ContentBased class(get_word_distribution). you can assume it has the size of (n_items, n_factors), where n_items is the number of articles according to the DataParser.get_abstracts function. So we can easily integrate it with the RecommenderSystem class.